src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,548 @@
+import json
+import subprocess
+import time
+
+import middlewared.sqlalchemy as sa
+from middlewared.plugins.jbof.redfish import (InvalidCredentialsError,
+                                              RedfishClient)
+from middlewared.schema import (Dict, Int, IPAddr, Password, Patch, Str,
+                                accepts, returns)
+from middlewared.service import CallError, CRUDService, ValidationErrors, private
+from middlewared.utils.license import LICENSE_ADDHW_MAPPING
+
+from .functions import (decode_static_ip, initiator_ip_from_jbof_static_ip,
+                        initiator_static_ip, jbof_static_ip,
+                        jbof_static_ip_from_initiator_ip,
+                        static_ip_netmask_int, static_ip_netmask_str,
+                        static_mtu)
+
+
+class JBOFModel(sa.Model):
+    __tablename__ = 'storage_jbof'
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+
+    jbof_description = sa.Column(sa.String(120), nullable=True)
+
+    # When performing static (code-based) assignment of data-plane IPs, we
+    # want each JBOD to have a deterministic unique index that counts up from
+    # zero (without any gaps, which rules out using the id).  This will not be
+    # part of the public API.
+    jbof_index = sa.Column(sa.Integer(), unique=True)
+    jbof_uuid = sa.Column(sa.Text(), nullable=False, unique=True)
+
+    jbof_mgmt_ip1 = sa.Column(sa.String(45), nullable=False)
+    jbof_mgmt_ip2 = sa.Column(sa.String(45), default='')
+    jbof_mgmt_username = sa.Column(sa.String(120))
+    jbof_mgmt_password = sa.Column(sa.EncryptedText())
+
+
+class JBOFService(CRUDService):
+
+    class Config:
+        service = 'jbof'
+        datastore = 'storage.jbof'
+        datastore_prefix = "jbof_"
+        cli_private = True
+        role_prefix = 'JBOF'
+
+    ENTRY = Dict(
+        'jbof_entry',
+        Int('id', required=True),
+
+        Str('description'),
+
+        # Redfish
+        IPAddr('mgmt_ip1', required=True),
```

Comment:
Are there any address types we want to exclude via `excluded_address_types=`?

Reply:
Debatable.  _Ideally_ we'd have a statically assigned IP address, but IMO if the shelf were configured with DHCP then a hostname tied to that would be preferable.